### ISSUE TITLE
IndentationCheck: ignoring line with sub-method call

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#3802]
````
$ cat TestClass.java
import java.nio.file.Path;

public class TestClass {
    void method(Path destinationPath, Path sourcePath) {
        Path destPath;
        destPath = destinationPath
.resolve(sourcePath.subpath(0));
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
    <module name="Indentation">
      <property name="basicOffset" value="4"/>
      <property name="braceAdjustment" value="0"/>
      <property name="caseIndent" value="4"/>
      <property name="throwsIndent" value="8"/>
    </module>
    </module>
</module>

$ java -jar checkstyle-7.5.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

Taken from https://github.com/checkstyle/contribution/pull/174#discussion_r99702559
We are expecting a violation on line 7 as it is not formatted correctly. Indentation is currently 0, we expect 12 to be the correct placement.

**Comment by kukreja-vikramaditya** [comment]
I am on it

**Comment by kukreja-vikramaditya** [comment]
The problem is occuring for all exprs with assignment operators and method calls 
`=` `+=` `-=` `* /=` `%=` `&=` `^=` `|=` `<<=` `>>=` `>>>=` 
```java
        int b = 1; //indent:8 exp:8 
        b += //indent:8 exp:8 
0; //indent:0 exp:12 warn 
        int c = 2; //indent:8 exp:8
        b = Integer //indent:8 exp:8 
.parseInt("1") +  c; //indent:0 exp:12 warn 

        b += Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b -= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b *= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b /= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b %= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b &= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b ^= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b |= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b <<= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b >>= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn

        b >>>= Integer //indent:8 exp:8
.parseInt("1") +  c; //indent:0 exp:12 warn
```

All warn lines except line 3 are ignored by CS. Should I create a new handler for such a specific case of assignment operators or add code to existing MethodCallHandler? Also I don't think MethodCallHandler is being notified of such cases. <- Could someone please verify this fact?

links - 
[MethodCallHandler](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/MethodCallHandler.java)
[IndentationCheck](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/IndentationCheck.java)

**Comment by rnveach** [comment]
@kukreja-vikramaditya I believe issue is with method call and not any type of assignment operator. We only validate tokens that start the line and the assignment operator doesn't start it in this issue.

> All warn lines except line 3 are ignored by CS. 

That shows it is the method call as line 3 doesn't have a method call.

I have not fully looked into the case, but looking at the AST tree of lines in question, issue could be related to the tree and could be similar to https://github.com/checkstyle/checkstyle/issues/3135#issuecomment-281365086 .
````
    |      |      `--METHOD_CALL -> ( [7:8]
    |      |          |--DOT -> . [7:0]
    |      |          |  |--IDENT -> destinationPath [6:19]
    |      |          |  `--IDENT -> resolve [7:1]
````

As noted in other issue, `resolve` is the first token on the line, but it is embedded inside `METHOD_CALL` and not in source order.

**Comment by kukreja-vikramaditya** [comment]
I removed the part `destPath = ` from  `destpath = destinationPath` but now CS is reporting the issue as expected. Would this have something to do with the `IDENT`?

ast when `IDENT` and `ASSIGN` are present
```
    |      |--EXPR -> EXPR [6:17]
    |      |  `--ASSIGN -> = [6:17]
    |      |      |--IDENT -> destPath [6:8]
    |      |      `--METHOD_CALL -> ( [7:8]
    |      |          |--DOT -> . [7:0]
    |      |          |  |--IDENT -> destinationPath [6:19]
    |      |          |  `--IDENT -> resolve [7:1]
 ```

ast without the ident and assignment
```
    |      |--EXPR -> EXPR [7:8]
    |      |  `--METHOD_CALL -> ( [7:8]
    |      |      |--DOT -> . [7:0]
    |      |      |  |--IDENT -> destinationPath [6:8]
    |      |      |  `--IDENT -> resolve [7:1]
  ``` 
I will go through the issue 3135 also as you suggested

**Comment by rnveach** [comment]
@kukreja-vikramaditya If 2 lines are similar but one is causing a violation and the other isn't and it should be, the first place you can start looking at is where the one with the violation is printing it out from in the code.

If we have code:
````
        string = ""//indent:8 exp:8
+ ""; //indent:0
````
violation is coming from `MethodDefHandler` with method `checkLineIndent` called by `checkLinesIndent` called by `checkExpressionSubtree` called by `checkChildren`.

If we have code:
````
        string = ""//indent:8 exp:8
.substring(0); //indent:0
````

Looking at method `checkExpressionSubtree`, the variable `subtreeLines` only has lines with `string = ""` and not `.substring(0)`. In the other example it had both lines. To me this seems like the issue.

I also stated similar findings in https://github.com/checkstyle/checkstyle/issues/3135#issuecomment-281365086 . So it does look like both issues are the same.

**Comment by kukreja-vikramaditya** [comment]
@rnveach 
Thanks for hint, I figured out that the problem is at [MethodCallHandler#checkIndentation](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/indentation/MethodCallHandler.java#L173) . 

In the current code base, it is assumed that `EXPR` is the parent node of `METHOD_CALL` (further whose parent is assumed to be `SLIST`) and the check is handled accordingly. But when we had added an `ASSIGN` operator, `METHOD_CALL`'s parent was `ASSIGN`, and its parent was `EXPR`. Not `SLIST` as the current code assumed. I can solve the problem specifically for `ASSIGN` but I think there could be more cases that create such a scenario, in which this has to be approached a little differently.

**Comment by kukreja-vikramaditya** [comment]
@romani 
@rnveach 
A little unrelated to the issue but just to clarify, what should be the indentation of L5 in this case ? 16?

```java
        Path destPath = destinationPath; //indent:8 exp:8
        destPath //indent:8 exp:8
            .getParent() //indent:12 exp:12
            .resolve(sourcePath //indent:12 exp:12
.subpath(0));  //indent:0
```

**Comment by rnveach** [comment]
> what should be the indentation of L5 in this case ? 

This sounds like it should in a separate issue. We try to go by what IDEs want as default as people will rely on them more than our check. IMO, it should be 16 since it is wrapping a wrapped line.

The other question seems more suitable for the other issue now.

**Comment by boyravikumar** [comment]
@kukreja-vikramaditya  whats the status of this.

**Comment by romani** [comment]
Discussion is started, but nobody is on it, everybody welcome to contribute

**Comment by romani** [comment]
@alinkov , if you know where how to fix this issue, we appreciate PR with fix.
